lib/cylc/network/httpserver.py

```diff
@@ -0,0 +1,853 @@
+            'tools.auth_digest.realm': self.suite,
+            'tools.auth_digest.get_ha1': self.get_ha1,
+            'tools.auth_digest.key': key,
+            'tools.auth_digest.algorithm': self.hash_algorithm
+        })
+        cherrypy.tools.connect_log = cherrypy.Tool(
+            'on_end_resource', self._report_connection_if_denied)
+        cherrypy.config['tools.connect_log.on'] = True
+        self.engine = cherrypy.engine
+        for port in self.ok_ports:
+            cherrypy.config["server.socket_port"] = port
+            try:
+                cherrypy.engine.start()
+                cherrypy.engine.wait(cherrypy.engine.states.STARTED)
+            except cherrypy.process.wspbus.ChannelFailures:
+                if cylc.flags.debug:
+                    traceback.print_exc()
+                # We need to reinitialise the httpserver for each port attempt.
+                cherrypy.server.httpserver = None
+            else:
+                if cherrypy.engine.state == cherrypy.engine.states.STARTED:
+                    self.port = port
+                    return
+        raise Exception("No available ports")
+
+    @staticmethod
+    def _get_client_connection_denied():
+        """Return whether a connection was denied."""
+        if "Authorization" not in cherrypy.request.headers:
+            # Probably just the initial HTTPS handshake.
+            return False
+        status = cherrypy.response.status
+        if isinstance(status, basestring):
+            return cherrypy.response.status.split()[0] in ["401", "403"]
+        return cherrypy.response.status in [401, 403]
+
+    def _report_connection_if_denied(self):
+        """Log an (un?)successful connection attempt."""
+        prog_name, user, host, uuid = _get_client_info()[1:]
+        connection_denied = self._get_client_connection_denied()
+        if connection_denied:
+            LOG.warning(self.__class__.LOG_CONNECT_DENIED_TMPL % (
+                user, host, prog_name, uuid))
+
+
+class SuiteRuntimeService(object):
+    """Suite runtime service API facade exposed via cherrypy."""
+
+    CLIENT_FORGET_SEC = 60
+    CLIENT_ID_MIN_REPORT_RATE = 1.0  # 1 Hz
+    CLIENT_ID_REPORT_SECONDS = 3600  # Report every 1 hour.
+    CONNECT_DENIED_PRIV_TMPL = (
+        "[client-connect] DENIED (privilege '%s' < '%s') %s@%s:%s %s")
+    LOG_COMMAND_TMPL = '[client-command] %s %s@%s:%s %s'
+    LOG_IDENTIFY_TMPL = '[client-identify] %d id requests in PT%dS'
+    LOG_FORGET_TMPL = '[client-forget] %s'
+    LOG_CONNECT_ALLOWED_TMPL = "[client-connect] %s@%s:%s privilege='%s' %s"
+
+    def __init__(self, schd):
+        self.schd = schd
+        self.clients = {}  # {uuid: time-of-last-connect}
+        self._id_start_time = time()  # Start of id requests measurement.
+        self._num_id_requests = 0  # Number of client id requests.
+
+    @cherrypy.expose
+    @cherrypy.tools.json_in()
+    @cherrypy.tools.json_out()
+    def clear_broadcast(
```

Comment:
Some thoughts on these methods:

- The privilege levels (e.g. `'full-control'`) could be constants.
- The `self._check_access_priv(<privlevel>)` method could be a decorator (e.g. `@privilege_full_control`)
- A `self._report()` call is present in every method? Could this be factored out?

Reply:
Added constants.

`self._check_access_priv(<privlevel>)`and `self._report()` now combined into a single method call. Not done via decorator, but it is now a single line per method.